Source/WebCore/platform/graphics/MIMESniffer.cpp

```diff
@@ -0,0 +1,341 @@
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "MIMESniffer.h"
+
+namespace WebCore {
+
+template<std::size_t N>
+constexpr auto span8(const char(&p)[N])
+{
+    return std::span<const uint8_t, N - 1>(reinterpret_cast<const uint8_t*>(p), N - 1);
+}
+
+static bool hasSignatureForMP4(std::span<const uint8_t> sequence)
+{
+    // To determine whether a byte sequence matches the signature for MP4, use the following steps:
+    // Let sequence be the byte sequence to be matched, where sequence[s] is byte s in sequence and sequence[0] is the first byte in sequence.
+    // Let length be the number of bytes in sequence.
+    size_t length = sequence.size();
+    // If length is less than 12, return false.
+    if (length < 12)
+        return false;
+    // Let box-size be the four bytes from sequence[0] to sequence[3], interpreted as a 32-bit unsigned big-endian integer.
+    uint32_t boxSize = sequence[0] << 24 | sequence[1] << 16 | sequence[2] << 8 | sequence[3] << 0;
+
+    // If length is less than box-size or if box-size modulo 4 is not equal to 0, return false.
+    if (length < boxSize || (boxSize % 4))
+        return false;
+    // If the four bytes from sequence[4] to sequence[7] are not equal to 0x66 0x74 0x79 0x70 ("ftyp"), return false.
+    if (sequence[4] != 0x66 || sequence[5] != 0x74 || sequence[6] != 0x79 || sequence[7] != 0x70)
+        return false;
+    // If the three bytes from sequence[8] to sequence[10] are equal to 0x6D 0x70 0x34 ("mp4"), return true.
+    if (sequence[8] == 0x6d && sequence[9] == 0x70 && sequence[10] == 0x34)
+        return true;
+
+    // Let bytes-read be 16.
+    uint32_t bytesRead = 16;
+
+    // This ignores the four bytes that correspond to the version number of the "major brand".
+    // While bytes-read is less than box-size, continuously loop through these steps:
+    while (bytesRead < boxSize) {
+        // If the three bytes from sequence[bytes-read] to sequence[bytes-read + 2] are equal to 0x6D 0x70 0x34 ("mp4"), return true.
+        if (sequence[bytesRead] == 0x6d && sequence[bytesRead + 1] == 0x70 && sequence[bytesRead + 2] == 0x34)
+            return true;
+        // Increment bytes-read by 4.
+        bytesRead += 4;
+    }
```

Comment:
Can we leverage ISOBox::peekBox() here?

Reply:
I wasn't aware of that method, however I really wanted us to be 100% spec compliant too and considering all the steps on how to read the box are defined in the spec; I feel it's better to follow it there.